src/code/c_keyframe.c

```diff
@@ -1,85 +1,1330 @@
+        result = FrameCtrl_UpdateOnce(frameCtrl);
+    } else {
+        result = FrameCtrl_UpdateLoop(frameCtrl);
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/c_keyframe/func_8018373C.s")
+    if (result == KEYFRAME_NOT_DONE) {
+        // Animation is not done, step curTime by (speed * (30.0f / 20.0f)), adjusting the sign if the animation is
+        // playing in reverse (end <= start)
+        speed = (frameCtrl->start < frameCtrl->end) ? frameCtrl->speed : -frameCtrl->speed;
+        frameCtrl->curTime = frameCtrl->curTime + speed * (30.0f / 20.0f);
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/c_keyframe/func_801837CC.s")
+    // Adjust time for looping
+    if (frameCtrl->curTime < 1.0f) {
+        // Wrap from the start to the end of the animation
+        frameCtrl->curTime = (frameCtrl->curTime - 1.0f) + frameCtrl->frameCount;
+    } else if (frameCtrl->frameCount < frameCtrl->curTime) {
+        // Wrap from the end to the start of the animation
+        frameCtrl->curTime = (frameCtrl->curTime - frameCtrl->frameCount) + 1.0f;
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/c_keyframe/func_80183808.s")
+    return result;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/c_keyframe/func_80183880.s")
+/**
+ * @note Original name unknown
+ */
+void Keyframe_ResetFlex(KFSkelAnimeFlex* kfSkelAnime) {
+    kfSkelAnime->skeleton = NULL;
+    kfSkelAnime->animation = NULL;
+    kfSkelAnime->jointTable = NULL;
+    kfSkelAnime->transformCallbacks = NULL;
+    kfSkelAnime->morphTable = NULL;
+    kfSkelAnime->morphFrames = 0.0f;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/c_keyframe/func_80183A3C.s")
+/**
+ * Initializes a flex-type keyframe skeleton. The initial animation type is KEYFRAME_ANIM_ONCE.
+ *
+ * @param skeleton Skeleton to animate
+ * @param animation Initial animation to use
+ * @param jointTable Joint table to store limb transformations. Should have enough space to store a root translation
+ *                   plus a limb rotation for all limbs in the skeleton.
+ * @param morphTable Joint table to store morph interpolation values. Should have enough space to store a root
+ *                   translation plus a limb rotation for all limbs in the skeleton.
+ * @param transformCallbacks Array of limb transformation callbacks that will be called when drawing a particular limb.
+ *                           The limb data contains the index to select which callback to run.
+ *
+ * @note Original name unknown
+ */
+void Keyframe_InitFlex(KFSkelAnimeFlex* kfSkelAnime, KeyFrameFlexSkeleton* skeleton, KeyFrameAnimation* animation,
+                       Vec3s* jointTable, Vec3s* morphTable, KeyframeTransformCallback* transformCallbacks) {
+    Keyframe_ResetFlex(kfSkelAnime);
+    FrameCtrl_Init(&kfSkelAnime->frameCtrl);
+    kfSkelAnime->skeleton = (KeyFrameFlexSkeleton*)Lib_SegmentedToVirtual(skeleton);
+    kfSkelAnime->animation = (KeyFrameAnimation*)Lib_SegmentedToVirtual(animation);
```

Comment:
This returns `void *` so there's no need for the cast here.
```suggestion
    kfSkelAnime->skeleton = Lib_SegmentedToVirtual(skeleton);
    kfSkelAnime->animation = Lib_SegmentedToVirtual(animation);
```

Could you check the other redundant casts?